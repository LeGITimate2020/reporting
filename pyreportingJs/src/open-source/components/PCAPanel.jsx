import React from 'react';
import PropTypes from 'prop-types';
import { zipObject, isEmpty } from 'lodash';
import ReactTable from 'react-table';
import SimpleBarChart from './SimpleBarChart';

/**
 * Pure React component to show Scree Plot
 */
const ScreePlot = ({ data }) => {
  const PCNames = data.map((item, index) => `PC${index + 1}`);
  const layout = {
    showlegend: false,
    width: 400,
    yaxis: {
      range: [0, 1],
    },
  };
  return (
    <div className="row">
      <h4>Scree Plot</h4>
      <div className="col-lg-1">
        {
          data.map((item, index) => (
            <div key={index}>{PCNames[index]}: {item.toFixed(3)}</div>
          ))
        }
      </div>
      <div className="col-lg-11">
        <SimpleBarChart xValues={PCNames} yValues={data} layout={layout} />
      </div>
    </div>
  );
};

ScreePlot.propTypes = {
  /**
   * Array of explained variance e.g. [0.85, 0.10, 0.05]
   * Must be in descending order
   * Should add up to 1.0 (100%)
   * */
  data: PropTypes.arrayOf(PropTypes.number).isRequired,
};

const Eigenvectors = ({ eigenvectors, columns }) => {
  const listOfObjects = eigenvectors.map(eigenvector => zipObject(columns, eigenvector));
  const reactTableColumns = columns.map(column => ({ Header: column, accessor: column }));
  const column = {
    Header: '',
    id: 'row',
    filterable: false,
    Cell: row => <div>PC{row.index + 1}</div>,
  };
  reactTableColumns.unshift(column);
  return (
    <div>
      <h4>Eigenvectors</h4>
      <ReactTable data={listOfObjects} columns={reactTableColumns} defaultPageSize={eigenvectors.length} />
    </div>
  );
};

Eigenvectors.propTypes = {

  /** e.g. [ [4, -1, 3.2], [0.32, -2, -7] ] */
  eigenvectors: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.number)).isRequired,

  /** e.g. ['petal width', 'petal height', 'color'] */
  columns: PropTypes.arrayOf(PropTypes.string).isRequired,
};

const PCAPanel = (props) => {
  if (isEmpty(props)) {
    return <div>Click the Run PCA button to populate this tab.</div>;
  }

  const {
    explainedVariance,
    eigenvectors,
    columns,
  } = props;

  return (
    <div>
      <h3>Principal Components</h3>
      <div>
        <ScreePlot data={explainedVariance} />
        <Eigenvectors eigenvectors={eigenvectors} columns={columns} />
      </div>
    </div>
  );
};

PCAPanel.propTypes = {
  /**
   * Array of explained variance e.g. [0.85, 0.10, 0.05]
   * Must be in descending order
   * */
  explainedVariance: PropTypes.arrayOf(PropTypes.number),

  /** e.g. [ [4, -1, 3.2], [0.32, -2, -7] ] */
  eigenvectors: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.number)),

  /** e.g. ['petal width', 'petal height', 'color'] */
  columns: PropTypes.arrayOf(PropTypes.string),
};

export default PCAPanel;
