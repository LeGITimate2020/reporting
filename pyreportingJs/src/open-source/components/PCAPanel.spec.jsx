import { shallow } from 'enzyme';
import React from 'react';
import PCAPanel from './PCAPanel';


describe('PCAPanel.jsx', () => {
  const props = {
    columns: ['col1', 'col2', 'col3'],
    eigenvectors: [[0.1, -0.1, -0.3], [0.3, 0.6, -0.1], [0.01, 0.02, 0.03]],
    explainedVariance: [0.80, 0.14, 0.06],
  };

  it('renders without crashing', () => {
    shallow(<PCAPanel {...props} />);
  });

  it('Return simple component with missing props', () => {
    const wrapper = shallow(<PCAPanel />);
    expect(wrapper.find('ScreePlot').length).toBe(0);
    expect(wrapper.find('Eigenvectors').length).toBe(0);
  });

  it('Return component with correct props', () => {
    const wrapper = shallow(<PCAPanel {...props} />);
    expect(wrapper.find('ScreePlot').length).toBe(1);
    expect(wrapper.find('Eigenvectors').length).toBe(1);
  });

  it('Matches snapshot without props', () => {
    const wrapper = shallow(<PCAPanel />);
    expect(wrapper).toMatchSnapshot();
  });

  it('Matches snapshot with props', () => {
    const wrapper = shallow(<PCAPanel {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
