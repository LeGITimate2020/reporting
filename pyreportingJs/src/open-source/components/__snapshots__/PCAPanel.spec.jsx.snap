// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PCAPanel.jsx Matches snapshot with props 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PCAPanel
    columns={
      Array [
        "col1",
        "col2",
        "col3",
      ]
    }
    eigenvectors={
      Array [
        Array [
          0.1,
          -0.1,
          -0.3,
        ],
        Array [
          0.3,
          0.6,
          -0.1,
        ],
        Array [
          0.01,
          0.02,
          0.03,
        ],
      ]
    }
    explainedVariance={
      Array [
        0.8,
        0.14,
        0.06,
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3>
          Principal Components
        </h3>,
        <div>
          <ScreePlot
            data={
              Array [
                0.8,
                0.14,
                0.06,
              ]
            }
          />
          <Eigenvectors
            columns={
              Array [
                "col1",
                "col2",
                "col3",
              ]
            }
            eigenvectors={
              Array [
                Array [
                  0.1,
                  -0.1,
                  -0.3,
                ],
                Array [
                  0.3,
                  0.6,
                  -0.1,
                ],
                Array [
                  0.01,
                  0.02,
                  0.03,
                ],
              ]
            }
          />
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Principal Components",
        },
        "ref": null,
        "rendered": "Principal Components",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <ScreePlot
              data={
                Array [
                  0.8,
                  0.14,
                  0.06,
                ]
              }
            />,
            <Eigenvectors
              columns={
                Array [
                  "col1",
                  "col2",
                  "col3",
                ]
              }
              eigenvectors={
                Array [
                  Array [
                    0.1,
                    -0.1,
                    -0.3,
                  ],
                  Array [
                    0.3,
                    0.6,
                    -0.1,
                  ],
                  Array [
                    0.01,
                    0.02,
                    0.03,
                  ],
                ]
              }
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "data": Array [
                0.8,
                0.14,
                0.06,
              ],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "columns": Array [
                "col1",
                "col2",
                "col3",
              ],
              "eigenvectors": Array [
                Array [
                  0.1,
                  -0.1,
                  -0.3,
                ],
                Array [
                  0.3,
                  0.6,
                  -0.1,
                ],
                Array [
                  0.01,
                  0.02,
                  0.03,
                ],
              ],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3>
            Principal Components
          </h3>,
          <div>
            <ScreePlot
              data={
                Array [
                  0.8,
                  0.14,
                  0.06,
                ]
              }
            />
            <Eigenvectors
              columns={
                Array [
                  "col1",
                  "col2",
                  "col3",
                ]
              }
              eigenvectors={
                Array [
                  Array [
                    0.1,
                    -0.1,
                    -0.3,
                  ],
                  Array [
                    0.3,
                    0.6,
                    -0.1,
                  ],
                  Array [
                    0.01,
                    0.02,
                    0.03,
                  ],
                ]
              }
            />
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Principal Components",
          },
          "ref": null,
          "rendered": "Principal Components",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <ScreePlot
                data={
                  Array [
                    0.8,
                    0.14,
                    0.06,
                  ]
                }
              />,
              <Eigenvectors
                columns={
                  Array [
                    "col1",
                    "col2",
                    "col3",
                  ]
                }
                eigenvectors={
                  Array [
                    Array [
                      0.1,
                      -0.1,
                      -0.3,
                    ],
                    Array [
                      0.3,
                      0.6,
                      -0.1,
                    ],
                    Array [
                      0.01,
                      0.02,
                      0.03,
                    ],
                  ]
                }
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "data": Array [
                  0.8,
                  0.14,
                  0.06,
                ],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "columns": Array [
                  "col1",
                  "col2",
                  "col3",
                ],
                "eigenvectors": Array [
                  Array [
                    0.1,
                    -0.1,
                    -0.3,
                  ],
                  Array [
                    0.3,
                    0.6,
                    -0.1,
                  ],
                  Array [
                    0.01,
                    0.02,
                    0.03,
                  ],
                ],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`PCAPanel.jsx Matches snapshot without props 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PCAPanel />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "Click the Run PCA button to populate this tab.",
    },
    "ref": null,
    "rendered": "Click the Run PCA button to populate this tab.",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Click the Run PCA button to populate this tab.",
      },
      "ref": null,
      "rendered": "Click the Run PCA button to populate this tab.",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
